File: sampleApp/src/main/java/com/zeyad/usecases/app/screens/user/detail/UserDetailVM.java
package com.zeyad.usecases.app.screens.user.detail;

import com.zeyad.usecases.api.IDataService;
import com.zeyad.usecases.app.components.redux.BaseEvent;
import com.zeyad.usecases.app.components.redux.BaseViewModel;
import com.zeyad.usecases.app.components.redux.SuccessStateAccumulator;
import com.zeyad.usecases.app.utils.Utils;
import com.zeyad.usecases.requests.GetRequest;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.functions.Function;

import static com.zeyad.usecases.app.utils.Constants.URLS.REPOSITORIES;

/**
 * @author zeyad on 1/10/17.
 */
public class UserDetailVM extends BaseViewModel<UserDetailState> {

    private IDataService dataUseCase;

    @Override
    public void init(SuccessStateAccumulator<UserDetailState> successStateAccumulator,
                     UserDetailState initialState, Object... otherDependencies) {
        setSuccessStateAccumulator(successStateAccumulator);
        setInitialState(initialState);
        dataUseCase = (IDataService) otherDependencies[0];
    }

    public Flowable<List<Repository>> getRepositories(String userLogin) {
        return Utils.isNotEmpty(userLogin) ? dataUseCase.<Repository>queryDisk(realm ->
                realm.where(Repository.class).equalTo("owner.login", userLogin))
                .flatMap(list -> Utils.isNotEmpty(list) ? Flowable.just(list) :
                        dataUseCase.<Repository>getList(new GetRequest.Builder(Repository.class, true)
                                .url(String.format(REPOSITORIES, userLogin))
                                .build())) :
                Flowable.error(new IllegalArgumentException("User name can not be empty"));
    }

    @Override
    public Function<BaseEvent, Flowable<?>> mapEventsToExecutables() {
        return event -> getRepositories(((GetReposEvent) event).getLogin());
    }
}

File: sampleApp/src/main/java/com/zeyad/usecases/app/screens/user/detail/UserDetailVM.kt
package com.zeyad.usecases.app.screens.user.detail

import com.zeyad.rxredux.core.BaseEvent
import com.zeyad.rxredux.core.viewmodel.BaseViewModel
import com.zeyad.rxredux.core.viewmodel.StateReducer
import com.zeyad.usecases.api.IDataService
import com.zeyad.usecases.app.utils.Constants.URLS.REPOSITORIES
import com.zeyad.usecases.requests.GetRequest
import io.reactivex.Flowable
import io.reactivex.functions.Function

/**
 * @author zeyad on 1/10/17.
 */
class UserDetailVM(private val dataUseCase: IDataService) : BaseViewModel<UserDetailState>() {

    override fun stateReducer(): StateReducer<UserDetailState> {
        return object : StateReducer<UserDetailState> {
            override fun reduce(newResult: Any, event: BaseEvent<*>, currentStateBundle: UserDetailState?): UserDetailState {
                return UserDetailState.builder()
                        .setRepos(newResult as List<Repository>).setUser(currentStateBundle?.user!!)
                        .setIsTwoPane(currentStateBundle.isTwoPane).build()
            }
        }
    }

    override fun mapEventsToActions(): Function<BaseEvent<*>, Flowable<*>> {
        return Function { event -> getRepositories((event as GetReposEvent).getPayLoad()) }
    }

    private fun getRepositories(userLogin: String): Flowable<List<Repository>> {
        return dataUseCase.getList(GetRequest
                .Builder(Repository::class.java, true)
                .url(String.format(REPOSITORIES, userLogin)).build())
//        return dataUseCase
//                .queryDisk(object : RealmQueryProvider<Repository> {
//                    override fun create(realm: Realm): RealmQuery<Repository> {
//                        return realm.where(Repository::class.java).equalTo("owner.login", userLogin)
//                    }
//                })
//                .flatMap { list ->
//                    if (!list.isEmpty())
//                        Flowable.just<List<Repository>>(list)
//                    else
//                        dataUseCase.getList(GetRequest
//                                .Builder(Repository::class.java, true)
//                                .url(String.format(REPOSITORIES, userLogin)).build())
//                }
    }
}
